### PR TITLE
[commonmark] make code fences compliant (+ fix 1058)

### LINKED ISSUE #1058: marked 0.3.14 hangs on the attached file, worked in 0.3.9
## Expectation

[example.txt](https://github.com/markedjs/marked/files/1735164/example.txt)
 should be converted to html. This worked with marked 0.3.9.

## Result

With marked 0.3.14 marked "hangs" when given the file.

## What was attempted

npm install marked@0.3.14
node_modules/marked/bin/marked

### LINKED ISSUE #1314: Whitespace in fenced code info string breaks rendering
**Describe the bug**

When there is whitespace in the info string of a fenced code block, like so: 
```
```html foo
```
...the block is not parsed as a code block. It appears to be rendered as a paragraph:
```html
<p>```html foo</p>
```

**To Reproduce**

1. Add a whitespace character in the info string of a fenced code block.
2. Render with the default renderer.

**Expected behavior**

When there is whitespace in the info string, GFM appears to use the first word as the language and ignore the rest. (As demonstrated [here](https://gist.github.com/camwiegert/0bfc9921993d22bf29b0b10a1ffda4ff)) I would expect Marked to do the same by default and pass the entire info string to a custom renderer.

Thank you! 👋

### LINKED ISSUE #1325: Nested pre render error
Here is a [demo](https://marked.js.org/demo/?text=%60%60%60%0A%2F%2F%60%60%60%0A%2F%2F%20const%20a%20%3D%20%27b%27%3B%0A%2F%2F%60%60%60%0A%60%60%60%0A%0A)

Should it be:
```
//```
// const a = 'b';
//```
```

Source md:
```md
    ```
    // ```
    // const a = 'b';
    // ```
    ```
```

### PR DESCRIPTION
**Pull request by Feder1co5oave**
Code fences containing a matching fence **not** on a new line are terminated incorrectly
- fixes #1058 ([comment](https://github.com/markedjs/marked/issues/1058#issuecomment-445346536))
- allow longer closing fences
- require indentation up to 3 spaces
- fixes #1325 
- fixes #1314

This gets up to 100% compliance with [commonmark fenced code blocks](https://spec.commonmark.org/0.28/#fenced-code-blocks)

- [x] Test(s) exist to ensure functionality and minimize regression (if no tests added, list tests covering this PR); or,

- [ ] Draft GitHub release notes have been updated.
- [ ] CI is green (no forced merge required).
- [ ] Merge PR

### CODE REVIEW

File: lib/marked.js [hunk 1]
```diff
@@ -920,7 +920,8 @@ function Renderer(options) {
   this.options = options || marked.defaults;
 }
 
-Renderer.prototype.code = function(code, lang, escaped) {
+Renderer.prototype.code = function(code, infostring, escaped) {
```
Comment (styfle) [inline lib/marked.js]:
```suggestion
Renderer.prototype.code = function(code, language, escaped) {
```

File: lib/marked.js [hunk 2]
```diff
@@ -920,7 +920,8 @@ function Renderer(options) {
...
+  var lang = (infostring || '').match(/\S*/)[0];
```
Comment (styfle) [inline lib/marked.js]:
```suggestion
  var lang = (language || '').match(/\S*/)[0];
```

**Comment by Feder1co5oave** [comment]
@styfle exactly, I changed the wording to better adhere to commonmark, see the paragraph and the examples below [example 110](https://spec.commonmark.org/0.28/#example-110). The language is the _first word_ of the info string, which may contain other information.

**Comment by Feder1co5oave** [comment]
Fix list updated.

File: lib/marked.js [hunk 3]
```diff
@@ -91,7 +91,7 @@ block.normal = merge({}, block);
  */
 
 block.gfm = merge({}, block.normal, {
-  fences: /^ *(`{3,}|~{3,})[ \.]*(\S+)? *\n([\s\S]*?)\n? *\1 *(?:\n+|$)/,
+  fences: /^ {0,3}(`{3,}|~{3,})([^`\n]*)\n(?:|([\s\S]*?)\n)(?: {0,3}\1[~`]* *(?:\n+|$)|$)/,
```
Comment (davisjam) [inline lib/marked.js]: This is no *more* vulnerable than it previously was.

**Pull request merged**